src/app/clusters/thread_network_diagnostics_server/thread_network_diagnostics_server.cpp

```diff
@@ -16,11 +16,116 @@
  */
 
 #include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
 #include <app/CommandHandler.h>
+#include <app/MessageDef/AttributeDataElement.h>
 #include <app/util/af.h>
+#include <app/util/attribute-storage.h>
+#include <lib/core/CHIPEncoding.h>
+#include <lib/core/CHIPTLVTypes.h>
+#include <lib/core/Optional.h>
+#include <lib/support/CHIPPlatformMemory.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConnectivityManager.h>
 
 using namespace chip;
+using namespace chip::app;
 using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::ThreadNetworkDiagnostics::Attributes;
+using namespace chip::DeviceLayer;
+
+namespace {
+
+class ThreadDiagosticsAttrAccess : public AttributeAccessInterface
+{
+public:
+    // Register for the EthernetNetworkDiagnostics cluster on all endpoints.
+    ThreadDiagosticsAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), ThreadNetworkDiagnostics::Id) {}
+
+    CHIP_ERROR Read(ClusterInfo & aClusterInfo, TLV::TLVWriter * aWriter, bool * aDataRead) override;
+
+private:
+    CHIP_ERROR ReadIfSupported(chip::AttributeId attributeId, TLV::TLVWriter * aWriter);
+};
+
+ThreadDiagosticsAttrAccess gAttrAccess;
+
+CHIP_ERROR ThreadDiagosticsAttrAccess::Read(ClusterInfo & aClusterInfo, TLV::TLVWriter * aWriter, bool * aDataRead)
+{
+    if (aClusterInfo.mClusterId != ThreadNetworkDiagnostics::Id)
+    {
+        // We shouldn't have been called at all.
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    *aDataRead = true;
+
+    return ReadIfSupported(aClusterInfo.mFieldId, aWriter);
```

Comment:
This does not seem right to me.  In particular, what happens if the client reads the ClusterRevision attribute?  Seems like this would error out for that.  Needs more test coverage?

Reply:
*aDataRead need to set to false if unexpected aClusterInfo.mFieldId is received